server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -77,65 +87,212 @@
+            nodeConnectionBlocks.put(node0, connectionBarrier::await);
+            transportService.disconnectFromNode(node0);
+
+            // can still connect to another node without blocking
+            final DiscoveryNode node1 = new DiscoveryNode("node1", buildNewFakeTransportAddress(), Version.CURRENT);
+            final DiscoveryNodes nodes1 = DiscoveryNodes.builder().add(node1).build();
+            final DiscoveryNodes nodes01 = DiscoveryNodes.builder(nodes0).add(node1).build();
+            final PlainActionFuture<Void> future1 = new PlainActionFuture<>();
+            service.connectToNodes(nodes01, () -> future1.onResponse(null));
+            future1.actionGet();
+            assertConnectedExactlyToNodes(nodes1);
+
+            // can also disconnect from node0 without blocking
+            final PlainActionFuture<Void> future2 = new PlainActionFuture<>();
+            service.connectToNodes(nodes1, () -> future2.onResponse(null));
+            future2.actionGet();
+            service.disconnectFromNodesExcept(nodes1);
+            assertConnectedExactlyToNodes(nodes1);
+
+            // however, now node0 is considered to be a new node so we will block on a subsequent attempt to connect to it
+            final PlainActionFuture<Void> future3 = new PlainActionFuture<>();
+            service.connectToNodes(nodes01, () -> future3.onResponse(null));
+            expectThrows(ElasticsearchTimeoutException.class, () -> future3.actionGet(timeValueMillis(scaledRandomIntBetween(1, 1000))));
+
+            // once the connection is unblocked we successfully connect to it.
+            nodeConnectionBlocks.clear();
+            connectionBarrier.await();
+            future3.actionGet();
+            assertConnectedExactlyToNodes(nodes01);
+
+            // if we disconnect from a node while blocked trying to connect to it then we do eventually disconnect from it
+            nodeConnectionBlocks.put(node0, connectionBarrier::await);
+            transportService.disconnectFromNode(node0);
+            final PlainActionFuture<Void> future4 = new PlainActionFuture<>();
+            service.connectToNodes(nodes01, () -> future4.onResponse(null));
+            future4.actionGet();
+            assertConnectedExactlyToNodes(nodes1);
+
+            service.disconnectFromNodesExcept(nodes1);
+            connectionBarrier.await();
+            if (randomBoolean()) {
+                // assertBusy because the connection completes before disconnecting, so we might briefly observe a connection to node0
+                assertBusy(() -> assertConnectedExactlyToNodes(nodes1));
+            }
+
+            // use ensureConnections() to wait until the service is idle
+            ensureConnections(service);
+            assertConnectedExactlyToNodes(nodes1);
+
+            // if we disconnect from a node while blocked trying to connect to it then the listener is notified
+            nodeConnectionBlocks.put(node0, connectionBarrier::await);
```

Comment:
This seems to already be in nodeConnectionBlocks here?

Reply:
Yes, 942ec1949dc32d4c8e5724d15c7975c8b9fd8fad